rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,122 @@
+
+#### Mapper Logic Inside IaSQL Engine
+
+Currently, we're developing our mappers in this category. But there might be other ways to integrate the logic inside the engine, without having to exactly follow our current pattern. I'm getting inspired by [Strapi](https://github.com/strapi/strapi), which is a headless CMS. You can create models using a GUI from Strapi's admin panel and the database will be changed on-air, new API routes will be added with proper logic, and it achieves these without losing performance or being in need of pushing new files to the codebase (but the codebase is changing on-air).
+
+So we can get inspired and create on-air mapper codes. But to ensure safety, we might take one of the two approaches: either by narrowing down the available variables, or the by narrowing down the available functionality.
+- Create a sandbox and run mapper's JavaScript code on IaSQL engine: in that sandbox, we can narrow down the context and not pass variables that might cause problems (not sure if that's technically possible currently, but I've found see some links: [1](https://nodejs.org/api/vm.html#vm_vm_runinthiscontext_code_options), [2](https://github.com/patriksimek/vm2)).
+  - The only issue that comes to my mind is that we will be vulnerable to SSRF attacks, which can be dangerous [in EC2](https://scalesec.com/blog/exploit-ssrf-to-gain-aws-credentials/). Our other vulnerability to SSRF would be related to the engine's Express server, which will be accessible by the running code. Therefore, we'll need authorization measures for that internal API.
+- Providing a new "safe" Domain Specific Language for IaSQL: developers can use this language to create new mappers, and we can translate that to TypeScript code on-air, while narrowing down the DSL interface to something that is "safe" (no code-execution, no unwanted database access, etc). The DSL does not have to be capable of creating all possible mappers (at least not at first) since if they want to do something very professional, they can just submit a PR to the engine.
+  - For example: we can supply a base provider like AWS. And using DSL people can dynamically invoke AWS client's methods and write their logic. The DSL then gets parsed and converted to JavaScript code, and we'll use its interface when we need the mapper.
+
+
+#### Mapper Logic Inside Postgres
+
+We can put the mapper's logic inside Postgres. And then provide an interface for the engine to call mapper functions. In this approach, the interface will be provided through the Postgres, and therefore this is the approach that is most consistent with "Pure-SQL". The DSL idea from the above way can still be applied here (developer writes DSL, PL/pgSQL code is generated).
+
+So let's consider one possible implementation of the mapper structure (all as Postgres functions):
+- CloudId: `{entity}_cloud_id(object: json) -> string` a Postgres function that takes the object as json and returns its cloud id as string.
```

Comment:
Should they be `json` types or Postgres `Row` types where the sub types match the table definition of the entity?

Reply:
This idea was also checked, and it's going to need an extra casting step more than this one.